File: vane-regions/src/main/java/org/oddlama/vane/regions/region/Role.java
package org.oddlama.vane.regions.region;

import static org.oddlama.vane.core.persistent.PersistentSerializer.fromJson;
import static org.oddlama.vane.core.persistent.PersistentSerializer.toJson;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;
import org.oddlama.vane.core.persistent.PersistentSerializer;

public class Role {

    public enum RoleType {
        ADMINS,
        OTHERS,
        NORMAL,
    }

    public static Object serialize(@NotNull final Object o) throws IOException {
        final var role = (Role) o;
        final var json = new JSONObject();
        json.put("id", PersistentSerializer.toJson(UUID.class, role.id));
        json.put("name", PersistentSerializer.toJson(String.class, role.name));
        json.put("roleType", PersistentSerializer.toJson(RoleType.class, role.roleType));
        try {
            json.put("settings", PersistentSerializer.toJson(Role.class.getDeclaredField("settings"), role.settings));
        } catch (NoSuchFieldException e) {
            throw new RuntimeException("Invalid field. This is a bug.", e);
        }

        return json;
    }

    @SuppressWarnings("unchecked")
    public static Role deserialize(@NotNull final Object o) throws IOException {
        final var json = (JSONObject) o;
        final var role = new Role();
        role.id = PersistentSerializer.fromJson(UUID.class, json.get("id"));
        role.name = PersistentSerializer.fromJson(String.class, json.get("name"));
        role.roleType = PersistentSerializer.fromJson(RoleType.class, json.get("roleType"));
        try {
            role.settings = (Map<RoleSetting, Boolean>) PersistentSerializer.fromJson(
                Role.class.getDeclaredField("settings"),
                json.get("settings")
            );
        } catch (NoSuchFieldException e) {
            throw new RuntimeException("Invalid field. This is a bug.", e);
        }
        return role;
    }

    private UUID id;
    private String name;
    private RoleType roleType;
    private Map<RoleSetting, Boolean> settings = new HashMap<>();

    private Role() {}

    public Role(final String name, final RoleType roleType) {
        this.id = UUID.randomUUID();
        this.name = name;
        this.roleType = roleType;
        for (final var rs : RoleSetting.values()) {
            this.settings.put(rs, rs.defaultValue(roleType == RoleType.ADMINS));
        }
    }

    public UUID id() {
        return id;
    }

    public String name() {
        return name;
    }

    public void name(final String name) {
        this.name = name;
    }

    public RoleType roleType() {
        return roleType;
    }

    public Map<RoleSetting, Boolean> settings() {
        return settings;
    }

    public boolean getSetting(final RoleSetting setting) {
        if (setting.hasOverride()) {
            return setting.getOverride() == 1;
        }
        return settings.getOrDefault(setting, setting.defaultValue(false));
    }

    public String color() {
        switch (roleType) {
            case ADMINS:
                return "§c";
            case OTHERS:
                return "§a";
            default:
            case NORMAL:
                return "§b";
        }
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/region/Role.kt
package org.oddlama.vane.regions.region

import org.json.JSONObject
import org.oddlama.vane.core.persistent.PersistentSerializer
import java.io.IOException
import java.util.*

/**
 * Role definition containing per-setting permissions inside a region group.
 */
class Role {
    /**
     * Built-in role categories with special semantics.
     */
    enum class RoleType {
        ADMINS,
        OTHERS,
        NORMAL,
    }

    /**
     * Unique role identifier.
     */
    private var id: UUID? = null
    /**
     * Role display name.
     */
    private var name: String? = null
    /**
     * Category of this role.
     */
    private var roleType: RoleType? = null
    /**
     * Explicit permission values configured for this role.
     */
    private var settings: MutableMap<RoleSetting?, Boolean?>? = EnumMap<RoleSetting, Boolean?>(RoleSetting::class.java)
    /**
     * Non-null accessor for the settings map.
     */
    private val settingsMap: MutableMap<RoleSetting?, Boolean?>
        get() = requireNotNull(settings)

    private constructor()

    /**
     * Creates a role with defaults based on its role type.
     */
    constructor(name: String?, roleType: RoleType?) {
        this.id = UUID.randomUUID()
        this.name = name
        this.roleType = roleType
        for (rs in RoleSetting.entries) {
            settingsMap[rs] = rs.defaultValue(roleType == RoleType.ADMINS)
        }
    }

    /**
     * Returns this role id.
     */
    fun id(): UUID? = id

    /**
     * Returns this role name.
     */
    fun name(): String? = name

    /**
     * Updates this role name.
     */
    fun name(name: String?) {
        this.name = name
    }

    /**
     * Returns this role category.
     */
    fun roleType(): RoleType? = roleType

    /**
     * Returns the mutable settings map.
     */
    fun settings(): MutableMap<RoleSetting?, Boolean?>? = settings

    /**
     * Returns the effective value for a role setting, including global overrides.
     */
    fun getSetting(setting: RoleSetting): Boolean {
        if (setting.hasOverride()) {
            return setting.override == 1
        }
        return settingsMap[setting] ?: setting.defaultValue(false)
    }

    /**
     * Returns the menu/chat color prefix for this role type.
     */
    fun color(): String = when (roleType) {
        RoleType.ADMINS -> "§c"
        RoleType.OTHERS -> "§a"
        RoleType.NORMAL -> "§b"
        else -> "§b"
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        /**
         * Serializes a role to JSON-compatible data.
         */
        fun serialize(o: Any): Any {
            val role = o as Role
            val json = JSONObject()
            putSerialized(json, "id", UUID::class.java, role.id)
            putSerialized(json, "name", String::class.java, role.name)
            putSerialized(json, "roleType", RoleType::class.java, role.roleType)
            withField {
                json.put(
                    "settings",
                    PersistentSerializer.toJson(Role::class.java.getDeclaredField("settings"), role.settings)
                )
            }

            return json
        }

        @JvmStatic
        @Throws(IOException::class)
        /**
         * Deserializes a role from JSON-compatible data.
         */
        fun deserialize(o: Any): Role {
            val json = o as JSONObject
            val role = Role()
            role.id = readSerialized(json, "id", UUID::class.java)
            role.name = readSerialized(json, "name", String::class.java)
            role.roleType = readSerialized(json, "roleType", RoleType::class.java)
            withField {
                role.settings = @Suppress("UNCHECKED_CAST")
                (PersistentSerializer.fromJson(
                    Role::class.java.getDeclaredField("settings"),
                    json.get("settings")
                ) as MutableMap<RoleSetting?, Boolean?>?)
            }
            return role
        }
    }
}
